src/NHibernate/Util/AsyncLock.cs

```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NHibernate.Util
+{
+	public class AsyncLock
```

Comment:
This `AsyncLock` implementation was taken from a MSDN blog post which was posted before .NET 4.5, and with it the method `SemaphoreSlim.WaitAsync()` was released. I think it would be better to use an `AsyncLock` implementation based on `SemaphoreSlim.WaitAsync()` like the one available here: https://www.hanselman.com/blog/ComparingTwoTechniquesInNETAsynchronousCoordinationPrimitives.aspx This would also mean that the new class `AsyncSemaphore` could be deleted again.